src/core/providers/ogr/qgsogrproviderutils.h

```diff
@@ -95,14 +95,15 @@ class CORE_EXPORT QgsOgrProviderUtils
 #else
         QRecursiveMutex mutex;
 #endif
+        QMutex sharedDSMutex;
         GDALDatasetH    hDS = nullptr;
         QMap<QString, QgsOgrLayer *>  setLayers;
         int            refCount = 0;
         bool           canBeShared = true;
 
         DatasetWithLayers()
 #if QT_VERSION < QT_VERSION_CHECK(5, 14, 0)
-          : mutex( QMutex::Recursive )
+          : mutex( QMutex::Recursive ), sharedDSMutex( QMutex::Recursive )
```

Comment:
if a new mutex object is indeed needed (which is not obvious, see previous comment), then if it is declared recursive in qt < 5.14, it should also be created as QRecursiveMutex for qt >= 5.14